rxsmartlock/src/main/java/com/freeletics/rxsmartlock/RxGoogleSmartLockManager.kt

```diff
@@ -32,73 +32,91 @@ object RxGoogleSmartLockManager : SmartLockManager {
 
     private val smartLockComponent = SmartLockComponent()
 
-    override fun retrieveCredentials(context: Context): Single<Credential> {
-        Timber.d("RetrieveCredentials started...")
-        startHiddenActivity(context)
-        return googleApiClientSubject
-                .firstOrError()
-                .flatMap {
-                    smartLockComponent.retrieveCredentialRequest(googleApiClient)
-                }
-                .doFinally { dispose() }
-    }
-
-    override fun storeCredentials(context: Context, credential: Credential): Completable {
-        Timber.d("storeCredentials started...")
-        startHiddenActivity(context)
-        return googleApiClientSubject
-                .firstOrError()
-                .flatMapCompletable {
-                    smartLockComponent.saveCredentialsRequest(googleApiClient, credential)
-                }
-                .doFinally { dispose() }
-    }
+    override fun retrieveCredentials(
+        context: Context
+    ): Single<Credential> = apiClient(context)
+        .doOnSubscribe {
+            Timber.d("RetrieveCredentials started...")
+        }
+        .flatMapSingle {
+            smartLockComponent.retrieveCredentialRequest(googleApiClient)
+        }
+        .firstOrError()
+
+    override fun storeCredentials(
+        context: Context,
+        credential: Credential
+    ): Completable = apiClient(context)
+        .doOnSubscribe {
+            Timber.d("storeCredentials started...")
+        }
+        .flatMap {
+            smartLockComponent
+                .saveCredentialsRequest(googleApiClient, credential)
+                .andThen(Observable.just(Unit))
+        }
+        .take(1)
+        .ignoreElements()
```

Comment:
What is the reason of doing here `.andThen(Observable.just(Unit))` and then `take(1).ignoreElements()`? Is it possible to use `flatMapCompletable()`?

Reply:
Special case here:
- `flatMapCompletable()` - waits for upstream for terminal event to finish itself and this will not happen with published observable as it never completes by itself.
- on the other hand `take(1)` closes upstream leading to disposing activity and Google api client while they still are used.

So to kind workaround this I have to implement this way.

Reply:
Got it, but this workaround can be confusing though